embassy-sync/src/blocking_mutex/mod.rs

```diff
@@ -3,109 +3,60 @@
-///
-/// Which implementation you select depends on the context in which you're using the mutex, and you can choose which kind
-/// of interior mutability fits your use case.
-///
-/// Use [`CriticalSectionMutex`] when data can be shared between threads and interrupts.
-///
-/// Use [`NoopMutex`] when data is only shared between tasks running on the same executor.
-///
-/// Use [`ThreadModeMutex`] when data is shared between tasks running on the same executor but you want a global singleton.
-///
-/// In all cases, the blocking mutex is intended to be short lived and not held across await points.
-/// Use the async [`Mutex`](crate::mutex::Mutex) if you need a lock that is held across await points.
-pub struct Mutex<R, T: ?Sized> {
-    // NOTE: `raw` must be FIRST, so when using ThreadModeMutex the "can't drop in non-thread-mode" gets
-    // to run BEFORE dropping `data`.
-    raw: R,
-    data: UnsafeCell<T>,
-}
-
-unsafe impl<R: RawMutex + Send, T: ?Sized + Send> Send for Mutex<R, T> {}
-unsafe impl<R: RawMutex + Sync, T: ?Sized + Send> Sync for Mutex<R, T> {}
-
-impl<R: RawMutex, T> Mutex<R, T> {
-    /// Creates a new mutex in an unlocked state ready for use.
-    #[inline]
-    pub const fn new(val: T) -> Mutex<R, T> {
-        Mutex {
-            raw: R::INIT,
-            data: UnsafeCell::new(val),
-        }
-    }
-
-    /// Creates a critical section and grants temporary access to the protected data.
-    pub fn lock<U>(&self, f: impl FnOnce(&T) -> U) -> U {
-        self.raw.lock(|| {
-            let ptr = self.data.get() as *const T;
-            let inner = unsafe { &*ptr };
-            f(inner)
-        })
-    }
-}
-
-impl<R, T> Mutex<R, T> {
-    /// Creates a new mutex based on a pre-existing raw mutex.
-    ///
-    /// This allows creating a mutex in a constant context on stable Rust.
-    #[inline]
-    pub const fn const_new(raw_mutex: R, val: T) -> Mutex<R, T> {
-        Mutex {
-            raw: raw_mutex,
-            data: UnsafeCell::new(val),
-        }
-    }
-
-    /// Consumes this mutex, returning the underlying data.
-    #[inline]
-    pub fn into_inner(self) -> T {
-        self.data.into_inner()
-    }
-
-    /// Returns a mutable reference to the underlying data.
-    ///
-    /// Since this call borrows the `Mutex` mutably, no actual locking needs to
-    /// take place---the mutable borrow statically guarantees no locks exist.
-    #[inline]
-    pub fn get_mut(&mut self) -> &mut T {
-        unsafe { &mut *self.data.get() }
-    }
-}
+use raw_mutex_traits::BlockingMutex;
+// Semver re-export
+pub use raw_mutex_traits::BlockingMutex as Mutex;
 
 /// A mutex that allows borrowing data across executors and interrupts.
 ///
 /// # Safety
 ///
 /// This mutex is safe to share between different executors and interrupts.
-pub type CriticalSectionMutex<T> = Mutex<raw::CriticalSectionRawMutex, T>;
+pub struct CriticalSectionMutex<T> {
```

Comment:
why change these from a type alias to a wrapper struct?

Reply:
We add some method definitions (namely `borrow`) in this file, since the `BlockingMutex` type is no longer defined in this crate, we can't add methods to it. So I made it a wrapper type so we could keep the same method defs on the named type